Improve docs (#116)
dcermak/pytest_container: source/tutorials.rst

Tutorials
=========

Getting started with ``pytest_container``
-----------------------------------------

.. note::

>    The following tutorial demonstrates how to get started using
>    ``pytest_container`` for a python project and to test the project's

Reviewer: ```suggestion
   The following tutorial demonstrates how to use
   ``pytest_container`` for a Python project and to test the project's
```

   Command-line interface.

The example uses  `poetry <https://python-poetry.org/>`_ to manage the Python
dependencies. If the :file:`pyproject.toml` doesn't already exist, create it using the :command:`poetry init`  command.

Start by adding ``pytest_container`` as a development dependency:

.. code-block:: shell-session

   ❯ poetry add --dev pytest_container

Add `pytest-xdist
<https://github.com/pytest-dev/pytest-xdist>`_, so that tests can be executed in
parallel:

.. code-block:: shell-session

   ❯ poetry add --dev pytest-xdist

> To keep things tidy, create a directory called :file:`tests`  for your tests. Add an empty file called :file:`tests/__init__.py`. Create the :file:`tests/conftest.py` file with the following contents:

Reviewer 2: "To keep things tidy" is a bit too abstract IMHO because it states an intention but not the effect. I would rephrase it as follows to put emphasis, that it's good to keep the tests atop of this framework in one single place, and that this place can be arbitrary. e.g.

```suggestion
It's highly recommended to create a new directory e.g. :file:`tests`  for your tests to keep things tidy. Add an empty file called :file:`tests/__init__.py`. Create the :file:`tests/conftest.py` file with the following contents:
```

Tutorials
=========

Getting started with ``pytest_container``
-----------------------------------------

.. note::

   In this tutorial we will start with a minimal python project and start using

>    ``pytest_container`` to test its command line interface.

Reviewer: _The following tutorial demonstrates how to_ would be better methinks.

- What does _its_ refer to?
- Command-line interface

> We will use `poetry <https://python-poetry.org/>`_ to manage our python
> dependencies in this example and assume that you already have a python project
> with a :file:`pyproject.toml`. If not, run :command:`poetry init` to create a
> new one interactively.

Reviewer: ```suggestion
The example uses  `poetry <https://python-poetry.org/>`_ to manage the Python
dependencies. If the :file:`pyproject.toml` doesn't already exist, create it using the :command:`poetry init`  command.
```

> First, we need to add ``pytest_container`` as a development dependency:

Reviewer: ```suggestion
Start by adding ``pytest_container`` as a development dependency:
```

.. code-block:: shell-session

   ❯ poetry add --dev pytest_container

> We also recommend to add `pytest-xdist
> <https://github.com/pytest-dev/pytest-xdist>`_ so that tests can be executed in
> parallel:

Reviewer: ```suggestion
Add `pytest-xdist
<https://github.com/pytest-dev/pytest-xdist>`_, so that tests can be executed in
parallel:
```

.. code-block:: shell-session

   ❯ poetry add --dev pytest-xdist

> Now we are ready to start writing our tests. Start by creating (if not yet
> present) a directory for your tests, preferably called :file:`tests` or
> :file:`test`. Add an empty :file:`tests/__init__.py` and create the file
> :file:`tests/conftest.py` with the following contents:

Reviewer: ```suggestion
To keep things tidy, create a directory called :file:`tests`  for your tests. Add an empty file called :file:`tests/__init__.py`. Create the :file:`tests/conftest.py` file with the following contents:
```

Author: :broom:

.. code-block:: python

   from pytest_container import auto_container_parametrize

   def pytest_generate_tests(metafunc):
       auto_container_parametrize(metafunc)

> The above code snippet ensures that the ``auto_container`` and
> ``auto_container_per_test`` fixtures work correctly (more specifically, that
> they are using the images defined in ``CONTAINER_IMAGES``).

Reviewer: ```suggestion
The above code snippet ensures that the ``auto_container`` and
``auto_container_per_test`` fixtures work correctly (that is,
they use the images defined in ``CONTAINER_IMAGES``).
```

We can now implement the actual tests. We will create a simple smoke test where
we install the python wheel of our project inside a container and then check
that the installed command line utility works as expected. For that, we will
start out with a :file:`Dockerfile`:

.. code-block:: Dockerfile

   FROM registry.opensuse.org/opensuse/tumbleweed

   RUN zypper -n in python3-pip
   COPY dist/*whl .
   RUN pip install *whl

> Assuming that this :file:`Dockerfile` were in our project's root and our python
> project is built into the :file:`dist/` folder, we would build an image based on
> openSUSE Tumbleweed with our project installed into it.

Reviewer: ```suggestion
Assuming that the :file:`Dockerfile` resides in the project's root and the Python
project is stored in the :file:`dist/` folder, build an image based on
openSUSE Tumbleweed with the project installed into it.
```

This still doesn't make sense to me. Project installed into what (what does _into it_ refer to)? Can project be installed into something?

> We can plug this :file:`Dockerfile` into ``pytest_container`` using the
> :py:class:`~pytest_container.container.DerivedContainer` class as follows:

Reviewer: ```suggestion
Plug the :file:`Dockerfile` into ``pytest_container`` using the
:py:class:`~pytest_container.container.DerivedContainer` class as follows:
```

.. code-block:: python

   from textwrap import dedent
   from pytest_container import DerivedContainer

   TW_WITH_PKG = DerivedContainer(
       base="registry.opensuse.org/opensuse/tumbleweed",
       containerfile=dedent("""
       RUN zypper -n in python3-pip
       COPY dist/*whl .
       RUN pip install *whl
       """)
   )

> Note that we omitted the ``FROM`` line from the
> :py:attr:`~pytest_container.container.DerivedContainer.containerfile` parameter
> as ``pytest_container`` will automatically include it for us.

Reviewer: ```suggestion
Note that  the ``FROM`` line is omitted from the
:py:attr:`~pytest_container.container.DerivedContainer.containerfile` parameter,
as ``pytest_container`` includes it automatically.
```

> Now we add the above snippet into a new file :file:`tests/test_cli.py` and add
> the following test function and global variable as well:

Reviewer: ```suggestion
Add the above snippet to a new file :file:`tests/test_cli.py`, and add
the following test function along with a global variable:
```

.. code-block:: python

   CONTAINER_IMAGES = [TW_WITH_PKG]

   def test_help_works(auto_container):
       res = auto_container.connection.run_expect([0], "my-binary --help")
       assert "My cool project" in res.stdout

> The global variable ``CONTAINER_IMAGES`` instructs ``pytest_container`` to run
> all test functions that use the ``auto_container`` or
> ``auto_container_per_test`` fixtures once for each image defined in that
> list. This allows you to have a single file with a plethora of tests that shall
> all be executed inside multiple container images and thus saves us the burden of
> manually parametrizing each of them.

Reviewer: ```suggestion
The global variable ``CONTAINER_IMAGES`` instructs ``pytest_container`` to run
all test functions that use the ``auto_container`` or
``auto_container_per_test`` fixtures once for each image defined in that
list. This allows you to have a single file with multiple of tests that to be executed inside multiple container images, thus avoiding the task of parametrizing each of the test manually.
```

> The test function receives a
> :py:class:`~pytest_container.container.ContainerData` instance, where we then use
> the :py:attr:`~pytest_container.container.ContainerData.connection` attribute to
> execute our projects binary. We leverage the `run_expect
> <https://testinfra.readthedocs.io/en/latest/modules.html#testinfra.host.Host.run_expect>`_
> function to execute the binary and check that its exit code is
> ``0``. Afterwards, we assert that a search string is in the standard output.

Reviewer: ```suggestion
The test function receives a
:py:class:`~pytest_container.container.ContainerData` instance, where 
the :py:attr:`~pytest_container.container.ContainerData.connection` attribute is used to
execute the project's binary. The `run_expect
<https://testinfra.readthedocs.io/en/latest/modules.html#testinfra.host.Host.run_expect>`_
function is used to execute the binary and check that its exit code is
``0``. Afterwards, we assert that a search string is in the standard output.
```
_leverage_ is a stupid word.

> Afterwards, we assert that a search string is in the standard output.

Assert a search  string? What does this mean?

> Now we can execute this test via :command:`poetry run pytest`.

Reviewer: ```suggestion
You can now execute this test via :command:`poetry run pytest`.
```
